sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/storage/common/storage_switch_to_secondary_policy.hpp>
+
+namespace Azure { namespace Storage { namespace Details {
+
+  std::unique_ptr<Azure::Core::Http::RawResponse> StorageSwitchToSecondaryPolicy::Send(
+      const Azure::Core::Context& ctx,
+      Azure::Core::Http::Request& request,
+      Azure::Core::Http::NextHttpPolicy nextHttpPolicy) const
+  {
+    bool considerSecondary = (request.GetMethod() == Azure::Core::Http::HttpMethod::Get
+                              || request.GetMethod() == Azure::Core::Http::HttpMethod::Head)
+        && !m_secondaryHost.empty();
+
+    int retryNumber = Azure::Core::Http::RetryPolicy::GetRetryNumber(ctx);
+    if (retryNumber > 0)
+    {
+      // switch host
+      if (request.GetUrl().GetHost() == m_primaryHost && considerSecondary)
+      {
+        request.GetUrl().SetHost(m_secondaryHost);
+      }
+      else
+      {
+        request.GetUrl().SetHost(m_primaryHost);
+      }
+    }
+
+    auto response = nextHttpPolicy.Send(ctx, request);
+
+    if (response->GetStatusCode() == Azure::Core::Http::HttpStatusCode::NotFound
+        || response->GetStatusCode() == Core::Http::HttpStatusCode::PreconditionFailed)
+    {
+      // FIXME: need a machenism to retain this value on a per-operation basis.
+      considerSecondary = false;
+    }
+
+    return response;
+  }
```

Comment:
Hi @JeffreyRichter , I ran into a problem when working on this retry-to-secondary policy.

if the resource is not yet replicated to secondary tenant, we need to disable retry-to-secondary mechanism. But there's no way to memorize this information in a policy.

The best place to save this information is in `Context`, but context is passed as const, so we cannot modify it.

Reply:
cc @RickWinter @vhvb1989 @ahsonkhan @antkmsft 

This is kind of urgent since it blocks the next release of core sdk and storage sdk

Reply:
I have a proposal.

We pass context in `HttpPolicy::Send` by non-const reference, so that information can be passed upwards along the policies. (Currently we can only pass information downwards). 
We keep context parameter in `HttpPipeline::Send` function const-reference unchanged and make a copy of context for each operation in `HttpPipeline::Send` function, so that concurrent operations don't interfere with each other.

Reply:
We can't pass the context back up because a single context can be shared by multiple threads and this would cause race conditions. I can think of 2 ways to "fix" this problem. You only need this feature for GET requests, right? So, in the Download method append a Context there whose value is a **pointer** to a "stop trying secondary" Boolean which you initialize to false. In the policy that looks for secondary, look up the key in the context and get the bool*. If false, check the secondary. If you check the secondary and fail, then set the *bool to true so that in the future for this GET, you will not check the secondary. The bool* Context will be destroyed after each GET operation. This "fix" you can easily implement today, right now.

Another option is we add a map to the HTTP request object. In the CheckSecondaryPolicy (or whatever you called it), look up the Boolean in the map. If not there, try the secondary. If you decide to stop trying the secondary, add the Boolean to the map. We don't have this map on HTTP request yet and the map is destroyed after each request.

Reply:
> Another option is we add a map to the HTTP request object.

This is basically the `RequestContext` struct. Azure Core decided not to create this because we already have the Context for this purpose.

@Jinming-Hu , The way we can solve this is a little ugly right now using the Context, but, basically, you need to:
1. Create a private Struct deriving from `Azure::ValueBase`  -> let's called it `StateForRetry`
2. Add a bool field to the struct
3. Create a per-operation-policy to run before the retry policy -> let's call it `StateForRetryPolicy`
4. Make this policy to create a new context with unique_ptr<StateForRetry>():
  ```cpp
    auto newContext = context.WithValue("retryOnSecondServer", std::make_unique<StateForRetry>());
    // Then use the newContext to call the next policy
   ```
5. Create two static method for `StateForRetryPolicy` called `IsSecondaryEnabled(context const&)` and `DisableRetryOnSecondary(context const&)`.
6. Implement these methods to look the key in the context and return the value or update it. 
6. Then, from the policy where you call Sencod server, Call `IsSecondatyEnabled()` first,  and if the resource is not replicated, call `DisavleRetryOnSecondaty()`

Check this PR: https://github.com/Azure/azure-sdk-for-cpp/pull/1796 as reference for how to access the unique_ptr value and update it.

Reply:
> @Jinming-Hu , The way we can solve this is a little ugly right now using the Context, but, basically, you need to:

> Create a private Struct deriving from Azure::ValueBase -> let's called it StateForRetry
Add a bool field to the struct
Create a per-operation-policy to run before the retry policy -> let's call it StateForRetryPolicy
Make this policy to create a new context with unique_ptr():
  auto newContext = context.WithValue("retryOnSecondServer", std::make_unique<StateForRetry>());
  // Then use the newContext to call the next policy
Create two static method for StateForRetryPolicy called IsSecondaryEnabled(context const&) and DisableRetryOnSecondary(context const&).
Implement these methods to look the key in the context and return the value or update it.
Then, from the policy where you call Sencod server, Call IsSecondatyEnabled() first, and if the resource is not replicated, call DisavleRetryOnSecondaty()
Check this PR: #1796 as reference for how to access the unique_ptr value and update it.

Thanks for the guide. Current implementation of Context is nasty to use.

Reply:
Done.

@JeffreyRichter @vhvb1989 If you want to take a look [storage_switch_to_secondary_policy.hpp](https://github.com/Azure/azure-sdk-for-cpp/blob/09cb4631dbcc61c300286280a1c10b7058e0dcc9/sdk/storage/azure-storage-common/inc/azure/storage/common/storage_switch_to_secondary_policy.hpp) [storage_switch_to_secondary_policy.cpp](https://github.com/Azure/azure-sdk-for-cpp/blob/09cb4631dbcc61c300286280a1c10b7058e0dcc9/sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp)

Reply:
If you  think about my suggestion in https://github.com/Azure/azure-sdk-for-cpp/pull/1745, 

adding a value to context can be as easy as

```C++
bool trySecondary = true;
context.WithValue("ReplicaStatusKey", &trySecondary);
```

and read the value

```C++
bool* trySecondary = context.Get<bool*>("ReplicaStatusKey");
```